File: scala/day1/part2.scala
package com.rjhallsted.adventOfCode2020.day1

object Part2 {
    def find3NumbersThatSumTo(goal: Int, numbers: Array[Int]): Option[(Int, Int, Int)] = {
        def inner(currentIndex: Int): Option[(Int, Int, Int)] = {
            if (currentIndex == numbers.length - 3) {
                return None
            }

            val subgoal = goal - numbers(currentIndex)
            (Part1.find2NumbersThatSumTo(subgoal, numbers, currentIndex + 1)) match {
                case (Some(x)) => Some((numbers(currentIndex), x._1, x._2))
                case (None) => inner(currentIndex + 1)
            }
        }

        inner(0)
    }
}

object Main2 extends App {
    val file = "../../inputs/day1.txt"
    val goal = 2020
    val inputs = Part1.getInputs(file).sorted

    (Part2.find3NumbersThatSumTo(goal, inputs)) match {
        case (Some(x)) => println(f"${x._1}, ${x._2}, ${x._3} -> ${x._1 * x._2 * x._3}") 
        case (None) => println("No solution")
    }
}

File: python/day1/part2.py
import functools

def getInput(filename):
    f = open(filename)
    numbers = f.readlines()
    f.close()
    return list(map(lambda x: int(x), numbers))

def find2NumbersThatSumTo(goal, numbers, start):
    end = len(numbers) - 1
    sum = numbers[start] + numbers[end]
    while sum != goal and end > start:
        if sum < goal:
            start += 1
        else:
            end -= 1
        sum = numbers[start] + numbers[end]
    if sum == goal:
        return (numbers[start], numbers[end])
    else:
        return False

def find3NumbersThatSumTo(goal, numbers, start):
    for i in range(len(numbers) - 2):
        subgoal = goal - numbers[i]
        pair = find2NumbersThatSumTo(subgoal, numbers, i + 1)
        if (pair != False):
            return (numbers[i], pair[0], pair[1])
    return False

def main():
    filename = "../../inputs/day1.txt"
    numbers = sorted(getInput(filename))
    result = find3NumbersThatSumTo(2020, numbers, 0)
    if (result == False):
        print("No result")
    else:
        print(*result)
        print(functools.reduce(lambda a,b: a*b, result))

main()